### PR TITLE
[JENKINS-54950] - use bindJSON(this,json) as default implementation for Descriptor#configure

### PR DESCRIPTION
**Pull request by ndeloof**
This change is intended to promote use of databinding relying on DataBoundSetters in Descriptors, as required by JCasC. 

See [JENKINS-54950](https://issues.jenkins-ci.org/browse/JENKINS-54950).

### Proposed changelog entries

* Internal: use bindJSON() as default implementation for Descriptor#configure

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). 
- [ ] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by jvz — APPROVED** [review]
The less code required for a plugin the better!

**Comment by daniel-beck** [comment]
So basically boilerplate code like https://github.com/jenkinsci/login-theme-plugin/blob/f9981980d1db118eaddae71f077df54e67589ac5/src/main/java/io/jenkins/plugins/logintheme/LoginTheme.java#L86...L91 will no longer be required. Nice.

**Comment by daniel-beck** [comment]
I'd be more comfortable with tests confirming this works as intended though TBH.

**Comment by ndeloof** [comment]
yes indeed. Make this the default, so people only override `configure` for some very specific corner cases, just like `newInstance` usually doesn't need any customisation by plugin maintainers

**Review by oleg-nenashev — APPROVED** [review]
LGTM. Some tests would be nice. Or you could just remove some obsolete code in Descriptor implementations within the core (a follow-up ticket for it would be nice)

**Comment by ndeloof** [comment]
@oleg-nenashev I indeed plan to review Descriptor implementations in core.
Main point I'd like to get review here is how to handle calls to super.configure, with risks for save() to be called twice. @jglick already recommended to rely on `BulkChange` for this purpose

**Comment by ndeloof** [comment]
I searched for a candidate Descriptor in jenkins-core I could use to write a testcase, but all those I've checked do some custom computation or adapt to legacy data structure.

**Comment by ndeloof** [comment]
@daniel-beck implemented a test case as hudson.model.DescriptorTest#descriptor_config_roudtrip
relying on (simplified) SCMTrigger.DescriptorImpl. 
As suggested by @oleg-nenashev some further descriptor cleanup could be proposed in a follow-up PR.

File: core/src/main/java/hudson/model/Descriptor.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Descriptor.java]: Add `@since TODO` if you want it to be a part of API. Otherwise makes sense to restrict it

File: core/src/main/java/hudson/model/Descriptor.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Descriptor.java]: IMO it should check `if (BulkChange.contains(this))` and invoke new BulkChange only when needed

**Review by oleg-nenashev — DISMISSED** [review]
Looks reasonable. IMO the transaction logic should be improved to run IFF there is no bending BulkChange for the object

File: core/src/main/java/hudson/model/Descriptor.java
Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: AFAIK BulkChange by design fully support being nested, doesn't it ?
BulkChange.contains is mostly used by `save` to bypass actual serialization on disk

Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: I can't find any other sample of using BulkChange with such an additional contains check

Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Descriptor.java]: I should review it more carefully

File: core/src/main/java/jenkins/tools/GlobalToolConfiguration.java
Comment (timja) [inline core/src/main/java/jenkins/tools/GlobalToolConfiguration.java]: BulkChange doesn't appear to be used here?

Comment (ndeloof) [reply core/src/main/java/jenkins/tools/GlobalToolConfiguration.java]: indeed, fixed

File: core/src/main/java/hudson/triggers/SCMTrigger.java
Comment (daniel-beck) [inline core/src/main/java/hudson/triggers/SCMTrigger.java]: Semantics change: This resets to default when invalid from user submission, the setter goes to the nearest boundary.

Also, don't we need a `@DataboundSetter` for this to work?

Comment (ndeloof) [reply core/src/main/java/hudson/triggers/SCMTrigger.java]: Semantic here indeed was heterogeneous. descriptor didn't used the same strategy when configured by webUI and direct setter invocation. 
I'd personally would prefer it reject improper values vs silently prefer a sane one without caller to know his proposed value was rejected, but this goes beyond this PR.

Comment (ndeloof) [reply core/src/main/java/hudson/triggers/SCMTrigger.java]: bindJSON(Object, json) do rely on javabean setters, no need for `@DataboundSetter` annotation - while this is required for Describable relying on bindJSON(class, json). Don't ask me about the logic here :)

File: core/src/main/java/hudson/model/Descriptor.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Descriptor.java]: `protected`?

File: core/src/main/java/hudson/model/Descriptor.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Descriptor.java]: The change drops the inline comment explaining (mostly, it ends with an unspecified core version) why this madness seems needed.

Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: I felt it was more confusing than helping in understanding what this is all about - at least it didn't helped me. I'm still not sure why this is needed.

Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]:
```
// collapse the structure to remain backward compatible with the JSON structure before 1.
``
`
this comment was introduced by https://github.com/jenkinsci/jenkins/commit/4ab6fe7cb4ccf2277da970447c50eab7c0921e47  (jenkins-1.494) but still doesn't make any sense to me with this context information.

Nothing get "collapsed" here afaict and this method just assumes target descriptor's json is nested within a larger json payload, anything else ? Maybe this comment was copy/pasted from another source ?

Comment (daniel-beck) [reply core/src/main/java/hudson/model/Descriptor.java]: Well, without any comment you look at it and wonder why this mutates the passed object etc. Seems pretty insane. So restoring the extended comment and pointing to the commit perhaps would be useful?

**Comment by ndeloof** [comment]
code review comments have been addressed, can we get this merged ?

File: core/src/main/java/hudson/model/Descriptor.java
Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: I also can't imagine any significant use-case to invoke `configure` within an existing BulkChange on the exact same object that would me this `contains` check relevant.

**Comment by ndeloof** [comment]
A major design issue I see here is that we don't give any solution to "reset to default before binding" issue, i.e is the json payload miss a value, the setter never will be executed (even with `null`) to reset field and give setter a change to use some default value.
https://github.com/stapler/stapler/pull/147 offers a possible alternative, but from the scope of this PR I can't find any reasonable way to ensure all setter get called just using the `bindJSON` approach.
Any suggestion ?

**Comment by daniel-beck** [comment]
JENKINS-21017 all over :(

**Comment by ndeloof** [comment]
this reminds me 2000's when I had to handle this same issue with Struts 1.0 forms :-\

**Comment by ndeloof** [comment]
I've added support for `bindJSON(bean,json)` in https://github.com/stapler/stapler/pull/147
This enforce all DataBoundSetters in a descriptor will get invoked either with data from json payload or with default value. Doing so we are guaranteed Descriptor will be fully (re)configured

**Comment by ndeloof** [comment]
I just noticed it might be nice we can _rollback_ changes when some FormException is thrown during data-binding. BulkChange doesn't offer this option, but I had in mind we could catch unexpected exception and invoke `load()` to restore Descriptor based on last successfully saved configuration. wdyt ?

File: core/src/main/java/jenkins/model/GlobalConfiguration.java
Comment (jglick) [inline core/src/main/java/jenkins/model/GlobalConfiguration.java]: Thus generalizing #2509 FTR.

**Comment by ndeloof** [comment]
`Could not find artifact org.kohsuke.stapler:stapler:jar:256-20181014.202408-2 in azure`
bu it is well deployed https://repo.jenkins-ci.org/snapshots/org/kohsuke/stapler/stapler/1.256-SNAPSHOT/stapler-1.256-20181014.202408-2.jar

**Comment by ndeloof** [comment]
testDeleteRecursive_onWindows FAILED
`java.lang.AssertionError: deletion stopped`

not sure about the cause and if/how this is related to my PR ...

**Comment by ndeloof** [comment]
@oleg-nenashev @jglick @batmat @daniel-beck you self-assigned for review, but didn't got much feedback then. How can we process forward on this ?

**Comment by jglick** [comment]
Indeed I have not had time to think about this.

File: core/src/main/java/hudson/model/Descriptor.java
Comment (ndeloof) [inline core/src/main/java/hudson/model/Descriptor.java]: I'm a bit worried we hide failure to save configuration here. Should throw a FormException. wdyt ?

**Review by alecharp — COMMENTED** [review]
I might not see the entire picture, but I couldn't see anything dramatic here.

File: core/src/main/java/hudson/model/Descriptor.java
Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: Agree, but FormException expect a specific field, and here the whole binding process did fail. Maybe better to throw an IOException ?
No concrete sample I can find in code to see the conventional way to address this

Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: Actually (sorry, this PR being 1 month old I can't remember such details) the original code forhudson.model.Descriptor#save doesn't send any feedback to end-user but just log failure to persist configuration, so I just did the same here.

File: core/pom.xml
Comment (alecharp) [reply core/pom.xml]: Perfect!

**Comment by oleg-nenashev** [comment]
Please merge the PR with upstream to rerun the tests

**Review by oleg-nenashev — COMMENTED** [review]
Also, IMHO it needs JIRA issue or link to any other document (like JEP) describing the change and dependencies on other stories

**Comment by oleg-nenashev** [comment]
https://github.com/stapler/stapler/pull/147 is still needed to merge that

**Comment by ndeloof** [comment]
created https://issues.jenkins-ci.org/browse/JENKINS-54950 for this specific PR

**Comment by jeffret-b** [comment]
I'm going to re-open to kick a build and see if the prior failure was just an infra issue.

File: core/src/main/java/hudson/Functions.java
Comment (Wadeck) [inline core/src/main/java/hudson/Functions.java]: Could you implement some unit tests for this method, especially cases for the catch statement due to the complexity ?

File: core/src/main/java/hudson/model/Descriptor.java
Comment (Wadeck) [inline core/src/main/java/hudson/model/Descriptor.java]: A test case to ensure this behavior is kept seems useful, WDYT ?

File: core/src/main/java/hudson/Functions.java
Comment (Wadeck) [inline core/src/main/java/hudson/Functions.java]: nit: + `f:number` + `f:textarea`

File: core/src/main/java/hudson/model/Descriptor.java
Comment (Wadeck) [reply core/src/main/java/hudson/model/Descriptor.java]: Change to consider 👍 as for all the new public methods in this PR

**Review by Wadeck — COMMENTED** [review]
Thank you @ndeloof for the concept, it's really interesting!

I will feel more confident with more automatic tests but also at least someone that tested it manually to ensure there is no other parts that are broken (or from plugins).

**Comment by ndeloof** [comment]
I'm closing this PR as I don't plan to invest more time on this, after months waiting for approval. Feel free to cherry-pick code if this topic is some way resurrected.

**Pull request closed without merging**
